### PR TITLE
add parTraverseN and parSequenceN

### PR DESCRIPTION
**Pull request by sideeffffect**
/cc @jendakol @augi

### CODE REVIEW

**Comment by sideeffffect** [comment]
Is there a way to work around the issue
```scala
[error]  * abstract method catsEffectSyntaxAsyncObj(cats.effect.Async)cats.effect.Async in trait cats.effect.syntax.AsyncSyntax is inherited by class AllCatsEffectSyntax in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("cats.effect.syntax.AllCatsEffectSyntax.catsEffectSyntaxAsyncObj")
[error]  * abstract method toAsyncOps(java.lang.Object,cats.effect.Async)cats.effect.Async#Ops in trait cats.effect.Async#ToAsyncOps is inherited by class AllCatsEffectSyntax in current version.
[error]    filter with: ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("cats.effect.syntax.AllCatsEffectSyntax.toAsyncOps")
[error] java.lang.RuntimeException: cats-effect: Binary compatibility check failed!
```
that happens only with Scala 2.11?

File: core/shared/src/main/scala/cats/effect/Async.scala [hunk 1]
```diff
@@ -319,6 +319,26 @@ object Async {
       }
     }
 
+  /**
+    * Like `Parallel.parTraverse`, but limits the degree of parallelism.
+    */
+  def parTraverseN[T[_]: Traverse, M[_], F[_], A, B](n: Long)(ta: T[A])(f: A => M[B])(implicit M: Async[M], P: Parallel[M, F]): M[T[B]] =
+    for {
+      semaphore <- Semaphore.uncancelable(n)(M)
```
Comment (SystemFw) [inline core/shared/src/main/scala/cats/effect/Async.scala]: Why are you using `uncancelable` here? That strikes me as a bug

Comment (sideeffffect) [reply core/shared/src/main/scala/cats/effect/Async.scala]: I can work with mere `Async` and don't need `Concurrent`. Would you suggest to change that to `Semaphore.apply`?

Comment (SystemFw) [reply core/shared/src/main/scala/cats/effect/Async.scala]: Yes

**Comment by sideeffffect** [comment]
@SystemFw any tips for solving the problem with Scala 2.11?

**Comment by djspiewak** [comment]
@sideeffffect It can't *really* be resolved on Scala 2.11. What you're doing here is you're adding functions to a trait (`AllCatsEffectSyntax`) which previously existed. Anyone inheriting from that trait is going to get that function. The problem with this is that, in Scala 2.11 (or more specifically, prior to Java 8), traits were encoded through explicit function delegation to a nested class (which contained the definition) in the implementing subtypes. This is problematic because the implementing subtypes *cannot* delegate to the nested class on these new functions (such as `parTraverseN`) since it didn't exist when they were compiled! (since they were compiled against previous versions of cats-effect) So any code which calls `parTraverseN` would then get an `AbstractMethodError`. This is what MiMa is detecting.

You basically cannot add syntax like this in the 1.x timeline. Or rather, you could add `AsyncSyntax`, but you can't add it to the main trait. The good news is that 2.x is right around the corner. The bad news is we really don't want to change *too* much, so I'm not sure if there's time for this to make it.

**Comment by rossabaker** [comment]
Is it any worse for compatibility than the changes proposed in #262?

**Comment by djspiewak** [comment]
@rossabaker Do you mean #519? Broadly-speaking no, it should be about the same if I fully comprehend both PRs.

**Comment by djspiewak** [comment]
@sideeffffect I would recommend waiting for #546 to land (which flushes out mima), then merge with master (please don't rebase open PRs) and see if that resolves the issues.

**Comment by djspiewak** [comment]
@sideeffffect Any update on this? We're going to be in compatibility-freeze very shortly.

**Comment by sideeffffect** [comment]
@djspiewak rebased on the newest master
OK to merge?

**Comment by djspiewak** [comment]
@sideeffffect Reminder: as I said, do not rebase PRs. It's in the contributing guidelines and everything. :-)

I'll take a final look at things and merge (hopefully later today). Thanks for all your work!

**Comment by sideeffffect** [comment]
@djspiewak *facepalm* sorry, I totally missed that
thanks for looking at it!

**Pull request merged**